File: src/class08/MINIPATH/Dijkstra/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;

// ========== 常量定义 ==========
static const int MAXN = 1e5+10; // 节点上限（题意中一般为10^5）
static const int MAXM = 2e5+10; // 边的上限（题意中一般为2×10^5）

// ========== 链式前向星存图相关数组 ==========
int head[MAXN], nxt[MAXM], to_[MAXM], w[MAXM];
int cnt; // 当前边下标（从 1 开始）

// ========== Dijkstra 需要的数组 ==========
int dist_[MAXN];   // dist_[u] 表示从起点 s 到 u 的最短距离
bool visited[MAXN]; // visited[u] 表示节点 u 是否已经确定最短距离

// ========== 输入数据 ==========
int n, m, s; // n: 节点数, m: 边数, s: 起点

// ========== 初始化 ==========
void build() {
    cnt = 1;
    for(int i = 1; i <= n; i++){
        head[i]    = 0;       // 没有边
        dist_[i]   = INT_MAX; // 最短距离初始为无穷
        visited[i] = false;   
    }
}

// ========== 添加边 ==========
// 在链式前向星中，为 u 添加一条有向边 (u -> v)，权重 wgt
void add_e(int u, int v, int wgt) {
    nxt[cnt]   = head[u];
    to_[cnt]   = v;
    w[cnt]     = wgt;
    head[u]    = cnt++;
}

// ========== 手写堆结构 ==========
// heap[i][0] 存节点编号, heap[i][1] 存该节点的距离
int heapArr[MAXM][2];
int hz; // 当前堆大小

// 交换堆元素
void swap_heap(int i, int j){
    // 交换节点编号
    int tmp = heapArr[i][0];
    heapArr[i][0] = heapArr[j][0];
    heapArr[j][0] = tmp;
    // 交换距离
    tmp = heapArr[i][1];
    heapArr[i][1] = heapArr[j][1];
    heapArr[j][1] = tmp;
}

// 上浮
void push_up(int i){
    while(i > 0 && heapArr[i][1] < heapArr[(i - 1) / 2][1]){
        swap_heap(i, (i - 1) / 2);
        i = (i - 1) / 2;
    }
}

// 下沉
void push_down(int i){
    int l=2*i+1;
    // 当 left 下标在堆内，说明有子节点
    while(l < hz){
        // 选出左右孩子中距离更小的
        int best = (l+1 < hz && heapArr[l+1][1] < heapArr[l][1]) 
                    ? l+1 
                    : l;
        //孩子和父亲比较
        best = heapArr[best][1]<heapArr[i][1]?best:i;
        //父亲优先级高， 那么终止下沉操作
        if(best==i){
            break;
        }
        swap_heap(best,i);
        i = best;
        l = 2*i+1;
    }
}

// ========== Dijkstra (懒更新写法) ==========
void dijkstra() {
    // 1) 起点距离设为 0
    dist_[s] = 0;

    // 2) 把起点 (节点 s, 距离 0) 放入堆
    heapArr[0][0] = s;
    heapArr[0][1] = 0;
    hz = 1; // 初始堆大小=1

    // 3) 循环
    while(hz > 0){
        // 取出堆顶
        int u = heapArr[0][0];

        // 把堆尾移到堆顶，堆大小-1
        swap_heap(0, --hz);
        push_down(0);

        // 判断是否已访问
        if(visited[u]) {
            continue;
        }
        visited[u] = true;

        // 松弛邻居
        for(int ei = head[u]; ei; ei = nxt[ei]) {
            int v = to_[ei];
            int wgt = w[ei];
            // 如果可以更新 dist_[v]
            if(!visited[v] && dist_[u] + wgt < dist_[v]) {
                dist_[v] = dist_[u] + wgt;
                // 入堆
                heapArr[hz][0] = v;
                heapArr[hz][1] = dist_[v];
                push_up(hz);
                hz++;
            }
        }
    }
}

// ========== 主函数 ==========
int main(){
    // 读入
    scanf("%d %d %d", &n, &m, &s);
    build();

    // 有向边输入
    for(int i = 0; i < m; i++){
        int u, v, wgt;
        scanf("%d %d %d", &u, &v, &wgt);
        // 只加一次 => 有向图
        add_e(u, v, wgt);
    }

    // Dijkstra
    dijkstra();

    // 输出 dist_[1]..dist_[n]
    // 注：题目若节点从1..n，这里一定要打印 dist_[1]
    printf("%d", dist_[1]);
    for(int i=2; i <= n; i++){
        printf(" %d", dist_[i]);
    }
    puts("");

    return 0;
}

File: src/class08/MINIPATH/Dijkstra/Dijkstra.java
import java.io.BufferedReader;
import java.io.StreamTokenizer;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.IOException;

//测试链接:https://www.luogu.com.cn/problem/P4779
//提交时修改主类名为Main
public class Dijkstra {
    public static final int MAXN = (int)1e5 + 10;
    public static final int MAXM = (int)2e5 + 10;

    // ========== 读取输入相关 ==========
    public static int n, m, s;             // n:节点数, m:边数, s:起点

    // ========== 链式前向星存图 ==========
    public static int[] head  = new int[MAXN];
    public static int[] next  = new int[MAXM];
    public static int[] to    = new int[MAXM];
    public static int[] weight = new int[MAXM];
    public static int cnt;

    // ========== Dijkstra 需要的辅助 ==========
    public static int[] dist   = new int[MAXN];
    public static boolean[] visited = new boolean[MAXN];

    // ========== 小顶堆结构 (heapArr[i][0] = 节点编号, heapArr[i][1] = 距离) ==========
    public static int[][] heapArr = new int[MAXM][2];
    public static int heapSize;

    // ========== 初始化 ==========
    public static void build() {
        cnt = 1;
        for(int i = 0; i <= n; i++){
            head[i] = 0;
            visited[i] = false;
            dist[i] = Integer.MAX_VALUE;
        }
        heapSize = 0;
    }

    // ========== 添加一条 (u->v) 的有向边，权重 wgt ==========
    public static void addEdge(int u, int v, int wgt){
        next[cnt]    = head[u];
        to[cnt]      = v;
        weight[cnt]  = wgt;
        head[u]      = cnt++;
    }

    // ========== 交换堆中两个位置 ==========
    public static void swap(int i,int j){
        int[] tmp   = heapArr[i];
        heapArr[i]  = heapArr[j];
        heapArr[j]  = tmp;
    }

    // ========== “上浮”操作：对下标 i 的元素进行向上调整 ==========
    public static void push_up(int i){
        while(heapArr[i][1] < heapArr[(i - 1) / 2][1]){
            swap(i, (i - 1) / 2);
            i = (i - 1) / 2;
        }
    }

    // ========== “下沉”操作：对下标 i 的元素进行向下调整 ==========
    //这里实际上仅对堆顶做下沉操作， 故而可以如下方式写
    public static void push_down(){
        int i = 0;
        int l = 1;
        while(l < heapSize){
            // 先比较左右孩子，选出更小的
            int r = l + 1;
            int best = (r < heapSize && heapArr[r][1] < heapArr[l][1])
                    ? r : l;
            // 再与父节点比较
            if(heapArr[best][1] < heapArr[i][1]){
                swap(best, i);
                i = best;
                l = i * 2 + 1;
            } else {
                break;
            }
        }
    }

    // ========== Dijkstra (懒更新写法) ==========
    public static void dijkstra(){
        // 1) 起点距离设为0，堆中插入 (s, 0)
        dist[s] = 0;
        heapArr[0][0] = s;
        heapArr[0][1] = 0;
        heapSize = 1;

        // 2) 主循环
        while(heapSize > 0){
            // 取出堆顶 (u, curDist)
            int u = heapArr[0][0];
            // 移除堆顶 => 将堆尾移到顶，再“下沉”
            swap(0, --heapSize);
            if(heapSize > 0) {
                push_down(0);
            }

            // 如果已访问过，跳过
            if(visited[u]) {
                continue;
            }
            // 标记 u 已确定最短距离
            visited[u] = true;

            // 3) 用 u 去松弛邻居 v
            for(int ei = head[u]; ei > 0; ei = next[ei]){
                int v = to[ei];
                int wgt = weight[ei];
                // 若可以松弛
                if(!visited[v] && dist[u] + wgt < dist[v]){
                    dist[v] = dist[u] + wgt;
                    // 插入到堆的末尾，令其“上浮”
                    heapArr[heapSize][0] = v;
                    heapArr[heapSize][1] = dist[v];
                    push_up(heapSize);
                    heapSize++;
                }
            }
        }
    }

    // ========== 主函数 ==========
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer in = new StreamTokenizer(br);
        PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));

        // 如果题目有多组测试，就用 while(in.nextToken() != ...) 包一层循环
        while(in.nextToken() != StreamTokenizer.TT_EOF){
            n = (int)in.nval;
            // 初始化
            build();

            in.nextToken(); m = (int)in.nval;
            in.nextToken(); s = (int)in.nval;

            // 读入 m 条边
            for(int i=0,u,v,wgt; i < m; i++){
                in.nextToken(); u   = (int)in.nval;
                in.nextToken(); v   = (int)in.nval;
                in.nextToken(); wgt = (int)in.nval;
                addEdge(u, v, wgt);
            }

            // 执行 Dijkstra
            dijkstra();

            // 输出 dist[1..n]
            out.print(dist[1]);
            for(int i = 2; i <= n; i++){
                out.print(" " + dist[i]);
            }
            out.println();
        }

        out.flush();
        br.close();
        out.close();
    }
}

